### PR TITLE
Fix reserving executable memory as per allocation type in ClrVirtualAllocWithinRange()

### LINKED ISSUE #83818: System.OutOfMemoryException when allocating jump stub on macOS ARM64
Hi!

Sometimes I get a `System.OutOfMemoryException` exception in .NET 7 app on macOS ARM64 in arbitrary places like:
```
System.OutOfMemoryException
at InvokeStub_ContextConsumersWithAttributesManager..ctor(Object, Object, IntPtr*)
at System.Reflection.ConstructorInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

Digging into this, it turned out that the exception is thrown from `HostCodeHeap::InitializeHeapList()`:
https://github.com/dotnet/runtime/blob/d037e070ebe5c83838443f869d5800752b0fcb13/src/coreclr/vm/dynamicmethod.cpp#L442-L449

because `ExecutableAllocator::Commit()` failed to commit executable memory:
https://github.com/dotnet/runtime/blob/d037e070ebe5c83838443f869d5800752b0fcb13/src/coreclr/vm/dynamicmethod.cpp#L758-L762

On macOS ARM64 `ExecutableAllocator::IsDoubleMappingEnabled()` is disabled, thus it internally calls `VirtualAlloc(..., MEM_COMMIT, PAGE_EXECUTE_READWRITE)` for the pre-reserved memory `HostCodeHeap::m_pLastAvailableCommittedAddr`, but then `mprotect(..., PROT_EXEC | PROT_READ | PROT_WRITE)` fails with errno `EACCES` (aka `Permission denied`):
https://github.com/dotnet/runtime/blob/d037e070ebe5c83838443f869d5800752b0fcb13/src/coreclr/pal/src/map/virtual.cpp#L1202-L1216

because initially this memory region was reserved w/o `MEM_RESERVE_EXECUTABLE` flag (which is used for `MAP_JIT`):
https://github.com/dotnet/runtime/blob/d037e070ebe5c83838443f869d5800752b0fcb13/src/coreclr/utilcode/util.cpp#L448-L460

despite the fact that `ExecutableAllocator::ReserveWithinRange()` actually requests it:
https://github.com/dotnet/runtime/blob/d037e070ebe5c83838443f869d5800752b0fcb13/src/coreclr/utilcode/executableallocator.cpp#L595-L605

Linked issues: https://github.com/dotnet/runtime/issues/50391, https://github.com/dotnet/runtime/pull/54954.

### PR DESCRIPTION
**Pull request by k15tfu**
Fixes #83818

### CODE REVIEW

File: src/coreclr/utilcode/util.cpp [hunk 1]
```diff
@@ -369,7 +369,13 @@ BYTE * ClrVirtualAllocWithinRange(const BYTE *pMinAddr,
     {
         NOTHROW;
         PRECONDITION(dwSize != 0);
-        PRECONDITION(flAllocationType == MEM_RESERVE);
+        PRECONDITION((flAllocationType & MEM_RESERVE) != 0);
+        PRECONDITION((flAllocationType & ~(
+                MEM_RESERVE
+#ifdef HOST_UNIX
```
Comment (janvorli) [inline src/coreclr/utilcode/util.cpp:375]: A nit -this doesn't do anything, since the precondition macros are empty for Unix.

Comment (janvorli) [reply src/coreclr/utilcode/util.cpp:375]: You also cannot have ifdef in there as it is inside of a macro expansion. I would just remove this and instead add a comment that on Unix, we can also have the `MEM_RESERVE_EXECUTABLE` set.

**Review by janvorli — COMMENTED** [review]
Great finding. It seems we have never discovered that as this means that all the preallocated memory from the executable allocator must be depleted before we hit this code path.

File: src/coreclr/utilcode/util.cpp (hunk 1 above)
Comment (k15tfu) [reply src/coreclr/utilcode/util.cpp]: Sure, does it look good now?

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Comment by k15tfu** [comment]
@janvorli Thank you!

Would it be possible to backport this to 7.0?

**Comment by janvorli** [comment]
@k15tfu I've created a porting PR and waiting for approval.

**Pull request merged**